src/com/google/javascript/jscomp/js/es6/array/copywithin.js

```diff
@@ -30,33 +34,41 @@ $jscomp.polyfill('Array.prototype.copyWithin', function(orig) {
    * @template VALUE
    */
   var polyfill = function(target, start, opt_end) {
+    if (this == null) {
+      throw new TypeError('this is null or not defined');
+    }
     var len = this.length;
-    target = Number(target);
-    start = Number(start);
-    opt_end = Number(opt_end != null ? opt_end : len);
-    if (target < start) {
-      opt_end = Math.min(opt_end, len);
-      while (start < opt_end) {
-        if (start in this) {
-          this[target++] = this[start++];
+    target = $jscomp.toInteger(target);
+    start = $jscomp.toInteger(start);
+    var end = opt_end === undefined ? len : $jscomp.toInteger(opt_end);
+    var to = target < 0 ? Math.max(len + target, 0) : Math.min(target, len);
```

Comment:
With 3 copies of Math.max and 4 copies of Math.min, I wonder if we should alias them?  I don't think the compiler is smart enough to know that this is allowed (since it can't tell that there aren't side effects on the receiver), so this may be a manual optimization that would help reduce code size a decent amount.

Alternatively, we could define the following function outside the closure:
```js
function clamp(len, value) {
  return value < 0 ? Math.max(0, value + len) : Math.min(value, len);
}
```
I don't know what the runtime cost of this would be (adding 3 non-native function calls to every invocation - given that this is running a potentially-long loop, that might be fine), but it would likely improve code size quite a bit: `var to = clamp(len, target); var from = clamp(len, start); ...`  The optimizer should be able to do a good job with that.

Reply:
@shicks 
Output
- original: [287 bytes gzipped (461 bytes uncompressed)](https://closure-compiler.appspot.com/home#code%3D%252F%252F%2520%253D%253DClosureCompiler%253D%253D%250A%252F%252F%2520%2540compilation_level%2520ADVANCED_OPTIMIZATIONS%250A%252F%252F%2520%2540output_file_name%2520default.js%250A%252F%252F%2520%253D%253D%252FClosureCompiler%253D%253D%250A%250A%2520%2520%252F**%250A%2520%2520%2520*%2520Copies%2520elements%2520from%2520one%2520part%2520of%2520the%2520array%2520to%2520another.%250A%2520%2520%2520*%250A%2520%2520%2520*%2520%2540this%2520%257B!IArrayLike%253CVALUE%253E%257D%250A%2520%2520%2520*%2520%2540param%2520%257Bnumber%257D%2520target%2520Start%2520index%2520to%2520copy%2520elements%2520to.%250A%2520%2520%2520*%2520%2540param%2520%257Bnumber%257D%2520start%2520Start%2520index%2520to%2520copy%2520elements%2520from.%250A%2520%2520%2520*%2520%2540param%2520%257Bnumber%253D%257D%2520opt_end%2520Index%2520from%2520which%2520to%2520end%2520copying.%250A%2520%2520%2520*%2520%2540return%2520%257B!IArrayLike%253CVALUE%253E%257D%2520The%2520array%252C%2520with%2520the%2520copy%2520performed%2520in-place.%250A%2520%2520%2520*%2520%2540template%2520VALUE%250A%2520%2520%2520*%252F%250A%2520%2520var%2520polyfill%2520%253D%2520function(target%252C%2520start%252C%2520opt_end)%2520%257B%250A%2520%2520%2520%2520var%2520len%2520%253D%2520this.length%253B%250A%2520%2520%2520%2520target%2520%253D%2520toInteger(target)%253B%250A%2520%2520%2520%2520start%2520%253D%2520toInteger(start)%253B%250A%2520%2520%2520%2520var%2520end%2520%253D%2520opt_end%2520%253D%253D%253D%2520undefined%2520%253F%2520len%2520%253A%2520toInteger(opt_end)%253B%250A%2520%2520%2520%2520var%2520to%2520%253D%2520target%2520%253C%25200%2520%253F%2520Math.max(len%2520%252B%2520target%252C%25200)%2520%253A%2520Math.min(target%252C%2520len)%253B%250A%2520%2520%2520%2520var%2520from%2520%253D%2520start%2520%253C%25200%2520%253F%2520Math.max(len%2520%252B%2520start%252C%25200)%2520%253A%2520Math.min(start%252C%2520len)%253B%250A%2520%2520%2520%2520var%2520final%2520%253D%2520end%2520%253C%25200%2520%253F%2520Math.max(len%2520%252B%2520end%252C%25200)%2520%253A%2520Math.min(end%252C%2520len)%253B%250A%2520%2520%2520%2520if%2520(to%2520%253C%2520from)%2520%257B%250A%2520%2520%2520%2520%2520%2520while%2520(from%2520%253C%2520final)%2520%257B%250A%2520%2520%2520%2520%2520%2520%2520%2520if%2520(from%2520in%2520this)%2520%257B%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520this%255Bto%252B%252B%255D%2520%253D%2520this%255Bfrom%252B%252B%255D%253B%250A%2520%2520%2520%2520%2520%2520%2520%2520%257D%2520else%2520%257B%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520delete%2520this%255Bto%252B%252B%255D%253B%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520from%252B%252B%253B%250A%2520%2520%2520%2520%2520%2520%2520%2520%257D%250A%2520%2520%2520%2520%2520%2520%257D%250A%2520%2520%2520%2520%257D%2520else%2520%257B%250A%2520%2520%2520%2520%2520%2520final%2520%253D%2520Math.min(final%252C%2520len%2520%252B%2520from%2520-%2520to)%253B%250A%2520%2520%2520%2520%2520%2520to%2520%252B%253D%2520final%2520-%2520from%253B%250A%2520%2520%2520%2520%2520%2520while%2520(final%2520%253E%2520from)%2520%257B%250A%2520%2520%2520%2520%2520%2520%2520%2520if%2520(--final%2520in%2520this)%2520%257B%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520this%255B--to%255D%2520%253D%2520this%255Bfinal%255D%253B%250A%2520%2520%2520%2520%2520%2520%2520%2520%257D%2520else%2520%257B%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520delete%2520this%255B--to%255D%253B%250A%2520%2520%2520%2520%2520%2520%2520%2520%257D%250A%2520%2520%2520%2520%2520%2520%257D%250A%2520%2520%2520%2520%257D%250A%2520%2520%2520%2520return%2520this%253B%250A%2520%2520%257D%253B%250A%250A%2520%2520%252F**%250A%2520%2520%2520*%2520%2540param%2520%257Bnumber%257D%2520arg%250A%2520%2520%2520*%2520%2540return%2520%257Bnumber%257D%250A%2520%2520%2520*%252F%250A%2520%2520function%2520toInteger(arg)%2520%257B%250A%2520%2520%2520%2520var%2520n%2520%253D%2520Number(arg)%253B%250A%2520%2520%2520%2520if%2520(n%2520%253D%253D%253D%2520Infinity%2520%257C%257C%2520n%2520%253D%253D%253D%2520-Infinity)%2520%257B%250A%2520%2520%2520%2520%2520%2520return%2520n%253B%250A%2520%2520%2520%2520%257D%250A%2520%2520%2520%2520return%2520n%2520%257C%25200%253B%250A%2520%2520%257D%253B%250A%250Aconsole.log(polyfill.apply(%255B0%252C%25201%252C%25202%252C%25203%252C%25204%252C%25205%255D%252C%2520%255BInfinity%252C%25200%255D))%253B)
- alias (`var max = Math.max`): [292 bytes gzipped (438 bytes uncompressed)](https://closure-compiler.appspot.com/home#code%3D%252F%252F%2520%253D%253DClosureCompiler%253D%253D%250A%252F%252F%2520%2540compilation_level%2520ADVANCED_OPTIMIZATIONS%250A%252F%252F%2520%2540output_file_name%2520default.js%250A%252F%252F%2520%253D%253D%252FClosureCompiler%253D%253D%250A%250A%2520%2520var%2520max%2520%253D%2520Math.max%253B%250A%2520%2520var%2520min%2520%253D%2520Math.min%253B%250A%250A%2520%2520%252F**%250A%2520%2520%2520*%2520Copies%2520elements%2520from%2520one%2520part%2520of%2520the%2520array%2520to%2520another.%250A%2520%2520%2520*%250A%2520%2520%2520*%2520%2540this%2520%257B!IArrayLike%253CVALUE%253E%257D%250A%2520%2520%2520*%2520%2540param%2520%257Bnumber%257D%2520target%2520Start%2520index%2520to%2520copy%2520elements%2520to.%250A%2520%2520%2520*%2520%2540param%2520%257Bnumber%257D%2520start%2520Start%2520index%2520to%2520copy%2520elements%2520from.%250A%2520%2520%2520*%2520%2540param%2520%257Bnumber%253D%257D%2520opt_end%2520Index%2520from%2520which%2520to%2520end%2520copying.%250A%2520%2520%2520*%2520%2540return%2520%257B!IArrayLike%253CVALUE%253E%257D%2520The%2520array%252C%2520with%2520the%2520copy%2520performed%2520in-place.%250A%2520%2520%2520*%2520%2540template%2520VALUE%250A%2520%2520%2520*%252F%250A%2520%2520var%2520polyfill%2520%253D%2520function(target%252C%2520start%252C%2520opt_end)%2520%257B%250A%2520%2520%2520%2520var%2520len%2520%253D%2520this.length%253B%250A%2520%2520%2520%2520target%2520%253D%2520toInteger(target)%253B%250A%2520%2520%2520%2520start%2520%253D%2520toInteger(start)%253B%250A%2520%2520%2520%2520var%2520end%2520%253D%2520opt_end%2520%253D%253D%253D%2520undefined%2520%253F%2520len%2520%253A%2520toInteger(opt_end)%253B%250A%2520%2520%2520%2520var%2520to%2520%253D%2520target%2520%253C%25200%2520%253F%2520max(len%2520%252B%2520target%252C%25200)%2520%253A%2520min(target%252C%2520len)%253B%250A%2520%2520%2520%2520var%2520from%2520%253D%2520start%2520%253C%25200%2520%253F%2520max(len%2520%252B%2520start%252C%25200)%2520%253A%2520min(start%252C%2520len)%253B%250A%2520%2520%2520%2520var%2520final%2520%253D%2520end%2520%253C%25200%2520%253F%2520max(len%2520%252B%2520end%252C%25200)%2520%253A%2520min(end%252C%2520len)%253B%250A%2520%2520%2520%2520if%2520(to%2520%253C%2520from)%2520%257B%250A%2520%2520%2520%2520%2520%2520while%2520(from%2520%253C%2520final)%2520%257B%250A%2520%2520%2520%2520%2520%2520%2520%2520if%2520(from%2520in%2520this)%2520%257B%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520this%255Bto%252B%252B%255D%2520%253D%2520this%255Bfrom%252B%252B%255D%253B%250A%2520%2520%2520%2520%2520%2520%2520%2520%257D%2520else%2520%257B%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520delete%2520this%255Bto%252B%252B%255D%253B%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520from%252B%252B%253B%250A%2520%2520%2520%2520%2520%2520%2520%2520%257D%250A%2520%2520%2520%2520%2520%2520%257D%250A%2520%2520%2520%2520%257D%2520else%2520%257B%250A%2520%2520%2520%2520%2520%2520final%2520%253D%2520min(final%252C%2520len%2520%252B%2520from%2520-%2520to)%253B%250A%2520%2520%2520%2520%2520%2520to%2520%252B%253D%2520final%2520-%2520from%253B%250A%2520%2520%2520%2520%2520%2520while%2520(final%2520%253E%2520from)%2520%257B%250A%2520%2520%2520%2520%2520%2520%2520%2520if%2520(--final%2520in%2520this)%2520%257B%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520this%255B--to%255D%2520%253D%2520this%255Bfinal%255D%253B%250A%2520%2520%2520%2520%2520%2520%2520%2520%257D%2520else%2520%257B%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520delete%2520this%255B--to%255D%253B%250A%2520%2520%2520%2520%2520%2520%2520%2520%257D%250A%2520%2520%2520%2520%2520%2520%257D%250A%2520%2520%2520%2520%257D%250A%2520%2520%2520%2520return%2520this%253B%250A%2520%2520%257D%253B%250A%250A%2520%2520%252F**%250A%2520%2520%2520*%2520%2540param%2520%257Bnumber%257D%2520arg%250A%2520%2520%2520*%2520%2540return%2520%257Bnumber%257D%250A%2520%2520%2520*%252F%250A%2520%2520function%2520toInteger(arg)%2520%257B%250A%2520%2520%2520%2520var%2520n%2520%253D%2520Number(arg)%253B%250A%2520%2520%2520%2520if%2520(n%2520%253D%253D%253D%2520Infinity%2520%257C%257C%2520n%2520%253D%253D%253D%2520-Infinity)%2520%257B%250A%2520%2520%2520%2520%2520%2520return%2520n%253B%250A%2520%2520%2520%2520%257D%250A%2520%2520%2520%2520return%2520n%2520%257C%25200%253B%250A%2520%2520%257D%253B%250A%250Aconsole.log(polyfill.apply(%255B0%252C%25201%252C%25202%252C%25203%252C%25204%252C%25205%255D%252C%2520%255BInfinity%252C%25200%255D))%253B)
- `clamp(len, value)`: [289 bytes gzipped (433 bytes uncompressed)](https://closure-compiler.appspot.com/home#code%3D%252F%252F%2520%253D%253DClosureCompiler%253D%253D%250A%252F%252F%2520%2540compilation_level%2520ADVANCED_OPTIMIZATIONS%250A%252F%252F%2520%2540output_file_name%2520default.js%250A%252F%252F%2520%253D%253D%252FClosureCompiler%253D%253D%250A%250A%2520%2520%252F**%250A%2520%2520%2520*%2520%2540param%2520%257Bnumber%257D%2520len%250A%2520%2520%2520*%2520%2540param%2520%257Bnumber%257D%2520value%250A%2520%2520%2520*%252F%250A%2520%2520function%2520clamp(len%252C%2520value)%2520%257B%250A%2520%2520%2520%2520return%2520value%2520%253C%25200%2520%253F%2520Math.max(0%252C%2520value%2520%252B%2520len)%2520%253A%2520Math.min(value%252C%2520len)%253B%250A%2520%2520%257D%250A%250A%2520%2520%252F**%250A%2520%2520%2520*%2520Copies%2520elements%2520from%2520one%2520part%2520of%2520the%2520array%2520to%2520another.%250A%2520%2520%2520*%250A%2520%2520%2520*%2520%2540this%2520%257B!IArrayLike%253CVALUE%253E%257D%250A%2520%2520%2520*%2520%2540param%2520%257Bnumber%257D%2520target%2520Start%2520index%2520to%2520copy%2520elements%2520to.%250A%2520%2520%2520*%2520%2540param%2520%257Bnumber%257D%2520start%2520Start%2520index%2520to%2520copy%2520elements%2520from.%250A%2520%2520%2520*%2520%2540param%2520%257Bnumber%253D%257D%2520opt_end%2520Index%2520from%2520which%2520to%2520end%2520copying.%250A%2520%2520%2520*%2520%2540return%2520%257B!IArrayLike%253CVALUE%253E%257D%2520The%2520array%252C%2520with%2520the%2520copy%2520performed%2520in-place.%250A%2520%2520%2520*%2520%2540template%2520VALUE%250A%2520%2520%2520*%252F%250A%2520%2520var%2520polyfill%2520%253D%2520function(target%252C%2520start%252C%2520opt_end)%2520%257B%250A%2520%2520%2520%2520var%2520len%2520%253D%2520this.length%253B%250A%2520%2520%2520%2520target%2520%253D%2520toInteger(target)%253B%250A%2520%2520%2520%2520start%2520%253D%2520toInteger(start)%253B%250A%2520%2520%2520%2520var%2520end%2520%253D%2520opt_end%2520%253D%253D%253D%2520undefined%2520%253F%2520len%2520%253A%2520toInteger(opt_end)%253B%250A%2520%2520%2520%2520var%2520to%2520%253D%2520clamp(len%252C%2520target)%253B%250A%2520%2520%2520%2520var%2520from%2520%253D%2520clamp(len%252C%2520start)%253B%250A%2520%2520%2520%2520var%2520final%2520%253D%2520clamp(len%252C%2520end)%253B%250A%2520%2520%2520%2520if%2520(to%2520%253C%2520from)%2520%257B%250A%2520%2520%2520%2520%2520%2520while%2520(from%2520%253C%2520final)%2520%257B%250A%2520%2520%2520%2520%2520%2520%2520%2520if%2520(from%2520in%2520this)%2520%257B%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520this%255Bto%252B%252B%255D%2520%253D%2520this%255Bfrom%252B%252B%255D%253B%250A%2520%2520%2520%2520%2520%2520%2520%2520%257D%2520else%2520%257B%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520delete%2520this%255Bto%252B%252B%255D%253B%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520from%252B%252B%253B%250A%2520%2520%2520%2520%2520%2520%2520%2520%257D%250A%2520%2520%2520%2520%2520%2520%257D%250A%2520%2520%2520%2520%257D%2520else%2520%257B%250A%2520%2520%2520%2520%2520%2520final%2520%253D%2520Math.min(final%252C%2520len%2520%252B%2520from%2520-%2520to)%253B%250A%2520%2520%2520%2520%2520%2520to%2520%252B%253D%2520final%2520-%2520from%253B%250A%2520%2520%2520%2520%2520%2520while%2520(final%2520%253E%2520from)%2520%257B%250A%2520%2520%2520%2520%2520%2520%2520%2520if%2520(--final%2520in%2520this)%2520%257B%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520this%255B--to%255D%2520%253D%2520this%255Bfinal%255D%253B%250A%2520%2520%2520%2520%2520%2520%2520%2520%257D%2520else%2520%257B%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520delete%2520this%255B--to%255D%253B%250A%2520%2520%2520%2520%2520%2520%2520%2520%257D%250A%2520%2520%2520%2520%2520%2520%257D%250A%2520%2520%2520%2520%257D%250A%2520%2520%2520%2520return%2520this%253B%250A%2520%2520%257D%253B%250A%250A%2520%2520%252F**%250A%2520%2520%2520*%2520%2540param%2520%257Bnumber%257D%2520arg%250A%2520%2520%2520*%2520%2540return%2520%257Bnumber%257D%250A%2520%2520%2520*%252F%250A%2520%2520function%2520toInteger(arg)%2520%257B%250A%2520%2520%2520%2520var%2520n%2520%253D%2520Number(arg)%253B%250A%2520%2520%2520%2520if%2520(n%2520%253D%253D%253D%2520Infinity%2520%257C%257C%2520n%2520%253D%253D%253D%2520-Infinity)%2520%257B%250A%2520%2520%2520%2520%2520%2520return%2520n%253B%250A%2520%2520%2520%2520%257D%250A%2520%2520%2520%2520return%2520n%2520%257C%25200%253B%250A%2520%2520%257D%253B%250A%250Aconsole.log(polyfill.apply(%255B0%252C%25201%252C%25202%252C%25203%252C%25204%252C%25205%255D%252C%2520%255BInfinity%252C%25200%255D))%253B)

Order (small is better)
- uncompressed size: clamp < alias < original
- gzipped size: original < clamp < alias
- function calls: original = alias < clamp

`Math.max` is probably called in other places, aliasing has little effect on gzipped size in real world (rather it may have a bad influence).
I think the original code has the best balance. What do you think?

Reply:
SGTM.